---
title: 
output:
  rmarkdown::html_document:
    theme: cerulean
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(root.dir = '/Users/boxiang/Documents/work/Baidu/projects/covseq/scripts/')
library(data.table)
library(ggplot2)
library(knitr)
library(rmarkdown)
library(DT)
library(cowplot)
library(lubridate)
library(sf)
library(rnaturalearth)
library(rnaturalearthdata)
library(countrycode)
```


<div class="jumbotron">
  <h2 class="display-3">Download CoV-Seq Datasets</h2>
  <hr class="my-10">
  <div class="row">
  
  <div class="col-sm-6">
  <p>Download SARS-CoV-2 variants and annotations</p>
  </div>
  
  <div class="col-sm-6">
  <p>Download aggregated metadata</p>
  </div>
  </div>
  
  <div class="row">
  
  <div class="col-sm-6">
  <a class="btn btn-primary btn-lg" href="https://covseq.s3-us-west-1.amazonaws.com/VCF/VCF.zip" role="button">Download VCF</a>
  </div>
  
  <div class="col-sm-6">
  <a class="btn btn-primary btn-lg" href="https://covseq.s3-us-west-1.amazonaws.com/metadata/metadata.zip" role="button">Download Metadata</a>
  </div>
  </div>
</div>

```{r metadata, echo=FALSE}
metadata = fread("../../data/aggregated/metadata/merged_in_vcf.tsv")
```


```{r display source, echo=FALSE}
data_source = metadata[, list(raw_data = .N), by="Data_Source"]
total = sum(data_source[,raw_data])
data_source = rbind(data_source, data.table(Data_Source="Total",raw_data=total))
```

## Browse All Sequences
```{r display metadata, echo=FALSE, warning = FALSE}
DT::datatable(metadata[,list(Accession_ID, Data_Source, Collection_Date, Country, Region)], filter = 'bottom', caption = 'Table 1: COVID-19 genomic sequence data source, collection date, and location.', colnames = c('Accession ID', 'Data Source', 'Collection Date', 'Country', "Region"))
```


## Number of sequence collected each week
```{r sequence vs time, echo=FALSE, warning=FALSE, fig.height=5, fig.width=12}
parsed_date = parse_date_time(metadata[,Collection_Date], c("y", "ym", "ymd"))
metadata$year_mon = sprintf("%s-%s",year(parsed_date), month(parsed_date))
metadata$week = week(parsed_date)
week_start = floor_date(parsed_date, unit="weeks")
week_end = ceiling_date(parsed_date, unit="weeks")
metadata$date_range = sprintf("%s to %s", week_start, week_end)
metadata$week_start = week_start
seq_vs_time = metadata[, list(num=.N), by=c("week_start","Data_Source")]
ggplot(seq_vs_time, aes(x=week_start,y=num,fill=Data_Source)) + geom_bar(stat="identity") + theme(axis.text.x = element_text(angle=0), legend.position = "bottom") + ylab("Number of Virus Collected/Sequenced") + xlab(NULL) + scale_fill_discrete(name="Source")
```


## Number of sequences by country of origin
```{r map, echo=FALSE, warning=FALSE, fig.height=5, fig.width=12}
world = ne_countries(scale="medium", returnclass="sf")
world$iso3c = countrycode(world$sovereignt, origin="country.name", destination = "iso3c")
metadata$iso3c = countrycode(metadata$Country, origin="country.name", destination = "iso3c")
seq_vs_ctry = metadata[, list(num_seq=.N), by="iso3c"]
to_plot = merge(world, seq_vs_ctry, by="iso3c", all.x=TRUE)
to_plot[is.na(to_plot$num_seq),]$num_seq = 0 
ggplot(data = to_plot) +
  theme_bw() + geom_sf(aes(fill=num_seq)) + 
  scale_fill_viridis_c(option = "plasma", trans = "sqrt", name = "No. Sequences")

```


## Look up variants annotations
```{r variant annotation, echo=FALSE, warning=FALSE}
annotation = fread("../../processed_data/snpEff/parse_snpEff/annotated.tsv")
DT::datatable(annotation[,list(POS, REF, ALT, EFFECT, IMPACT, GENE,HGVS_C,HGVS_P, CDS_POS, AA_POS, DISTANCE)], filter = 'bottom', caption = 'Table 2: Genomic Variant Annotation from SnpEff.', colnames = c("Pos", "Ref", "Alt", "Effect", "Impact", "Gene","HGVS.C", "HGVS.P", "CDS position", "AA position", "Distance"))
```


