---
title: "Test"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(root.dir = '/Users/boxiang/Documents/work/Baidu/projects/viraviz/scripts/')
library(data.table)
library(ggplot2)
```

```{r remove_duplicates}
in_fn = "../../processed_data/preprocess/filter_fasta/duplicated.tsv"
duplicated = fread(in_fn, sep="\t")
sum(duplicated$duplicated == 1)
```

```{r}
in_fn = "../../processed_data/preprocess/filter_fasta/genome_lengths.tsv"
genome_lengths = fread(in_fn, sep="\t", col.names = c("description", "length"))
p = ggplot(genome_lengths, aes(x=length)) + geom_histogram(bins=100) + scale_y_sqrt() + geom_vline(xintercept = 25000, color="red", linetype="dashed") + geom_text(x = 25000, y = 30, label = "Cutoff = 25000", color="red", hjust=1.1, size = 5)
print(p)
```


```{r ambiguous bases}
in_fn = "../../processed_data/preprocess/filter_fasta/ambiguous_bases.tsv"
ambiguous_bases = fread(in_fn, sep="\t", col.names = c("description", "count", "length", "proportion"))
p = ggplot(ambiguous_bases, aes(x=proportion + 1e-5)) + geom_histogram(bins=1000) + scale_x_log10()
print(p)
```


```{r mutation count}
in_fn = "../../processed_data/vcf/filter_samples/mutation_count.tsv"
mutation_count = fread(in_fn, sep="\t", col.names=c("sample", "mutation"))
p = ggplot(mutation_count, aes(x=mutation)) + geom_histogram(bins=300) + scale_y_sqrt() + scale_x_log10(breaks=c(1,10,100,150,1000,10000)) + geom_vline(xintercept=150, color="red", linetype="dashed")
print(p)
```




